frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -303,17 +326,43 @@ export default ({
         }
       })
     },
-    setInitMessages () {
-      if (this.isJoinedChatRoom(this.currentChatRoomId)) {
-        this.messages = cloneDeep(this.chatRoomLatestMessages)
+    getSimulatedState (initialize = true) {
+      return {
+        settings: cloneDeep(this.chatRoomSettings),
+        attributes: cloneDeep(this.chatRoomAttributes),
+        users: cloneDeep(this.chatRoomUsers),
+        messages: initialize ? [] : this.messages,
+        simulation: true
+      }
+    },
+    async getLatestEvents (refresh = false) {
+      const eventsPerPage = this.chatRoomSettings?.actionsPerPage || CHATROOM_ACTIONS_PER_PAGE
+      const beforeActionHash = refresh || !this.latestEvents.length
+        ? ''
+        : GIMessage.deserialize(this.latestEvents[0]).hash()
+
+      const newEvents = await sbp('chelonia/contractEventsBefore', this.currentChatRoomId, beforeActionHash, eventsPerPage)
+
+      if (refresh) {
+        this.latestEvents = cloneDeep(newEvents)
       } else {
-        this.messages = []
-        sbp('okTurtles.events/once', `${CHATROOM_STATE_LOADED}-${this.currentChatRoomId}`, (state) => {
-          this.messages = getLatestMessages({
-            count: this.chatRoomSettings.messagesPerPage, // TODO: this.chatRoomSettings could be {}
-            messages: state.messages
-          })
-        })
+        this.latestEvents = newEvents.concat(this.latestEvents)
```

Comment:
As a suggestion, `this.latestEvents.splice(0, newEvents.length, ...newEvents)` might avoid old `this.latestEvents` needing garbage-collection, without losing reactivity.

Reply:
Is it `this.latestEvents.splice(0, 0, ...newEvents)`?

Reply:
Hmm yes, sorry ! 😅

Reply:
Thanks. Done.